control-statistics/src/main/java/fi/nls/oskari/control/statistics/user/AddIndicatorDataHandler.java

```diff
@@ -0,0 +1,109 @@
+package fi.nls.oskari.control.statistics.user;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import fi.nls.oskari.annotation.OskariActionRoute;
+import fi.nls.oskari.control.*;
+import fi.nls.oskari.control.statistics.StatisticsHelper;
+import fi.nls.oskari.control.statistics.data.IndicatorValue;
+import fi.nls.oskari.control.statistics.data.IndicatorValueFloat;
+import fi.nls.oskari.control.statistics.data.StatisticalIndicator;
+import fi.nls.oskari.control.statistics.data.StatisticalIndicatorDataModel;
+import fi.nls.oskari.control.statistics.plugins.StatisticalDatasourcePlugin;
+import fi.nls.oskari.control.statistics.plugins.StatisticalDatasourcePluginManager;
+import fi.nls.oskari.control.statistics.plugins.db.DatasourceLayer;
+import fi.nls.oskari.service.OskariComponentManager;
+import fi.nls.oskari.util.JSONHelper;
+import fi.nls.oskari.util.ResponseHelper;
+import org.json.JSONObject;
+import org.oskari.statistics.user.StatisticalIndicatorService;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@OskariActionRoute("AddIndicatorData")
+public class AddIndicatorDataHandler extends ActionHandler {
+
+    private StatisticalIndicatorService indicatorService;
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    @Override
+    public void init() {
+        super.init();
+        if (indicatorService == null) {
+            indicatorService = OskariComponentManager.getComponentOfType(StatisticalIndicatorService.class);
+        }
+    }
+
+    private static String PARAM_DATA = "data";
+
+    public void handleAction(ActionParameters params) throws ActionException {
+        params.requireLoggedInUser();
+
+        int datasourceId = params.getRequiredParamInt(StatisticsHelper.PARAM_DATASOURCE_ID);
+        StatisticalDatasourcePlugin datasource = StatisticalDatasourcePluginManager.getInstance().getPlugin(datasourceId);
+        if(datasource == null) {
+            throw new ActionParamsException("Invalid datasource:" + datasourceId);
+        }
+        if(!datasource.canModify(params.getUser())) {
+            throw new ActionDeniedException("User doesn't have permission to modify datasource:" + datasourceId);
+        }
+        String id = params.getRequiredParam(ActionConstants.PARAM_ID);
```

Comment:
Since these parameters are required it would probably be better if we checked for their presence and correctness right at the beginning of the function instead. That way we don't waste processing on requests that are bound to fail.